contracts/crowdsale/ZxcCrowdsale.sol

```diff
@@ -0,0 +1,275 @@
+   * @param _rate ZXC/ETH exchange rate.
+   * @param _crowdSaleZxcCap Number of ZXC tokens offered during the sale.
+   * @param _bonusPresale Bonus token percentage for presale.
+   * @param _bonusSale Bonus token percentage for public sale stage with bonus.
+   * @param _minimumWeiDeposit Minimum required deposit in wei.
+   */
+  constructor(address _walletAddress,
+    address _tokenAddress,
+    uint256 _startTimePresale,  // 1529971200: date -d '2018-06-26 00:00:00 UTC' +%s
+    uint256 _startTimeSaleWithBonus, // 1530662400: date -d '2018-07-04 00:00:00 UTC' +%s
+    uint256 _startTimeSaleNoBonus,  //1530748800: date -d '2018-07-05 00:00:00 UTC' +%s
+    uint256 _endTime,  // 1531872000: date -d '2018-07-18 00:00:00 UTC' +%s
+    uint256 _rate,  // 10000: 1 ETH = 10,000 ZXC
+    uint256 _crowdSaleZxcCap, // 250M
+    uint256 _bonusPresale,  // 10 (%)
+    uint256 _bonusSale,  // 5 (%)
+    uint256 _minimumWeiDeposit  // 1 ether;
+  )
+    public
+  {
+    require(_walletAddress != address(0));
+    require(_tokenAddress != address(0));
+    require(_tokenAddress != _walletAddress);
+
+    token = Zxc(_tokenAddress);
+
+    uint8 _tokenDecimals = token.decimals();
+    require(_tokenDecimals == 18);  // Sanity check.
+    wallet = _walletAddress;
+
+    // Bonus should be > 0% and <= 100%
+    require(_bonusPresale > 0 && _bonusPresale <= 100);
+    require(_bonusSale > 0 && _bonusSale <= 100);
+    // 100% / _bonusPresale = bonusPresale -> bonus: tokenAmount / bonusPresale
+    bonusPresale = _bonusPresale;
+    // 100% / _bonusSale = bonusSale -> bonus: tokenAmount / bonusSale
+    bonusSale = _bonusSale;
+
+    require(_startTimePresale >= now);
+    require(_startTimeSaleWithBonus > _startTimePresale);
+    require(_startTimeSaleNoBonus > _startTimeSaleWithBonus);
+
+    startTimePresale = _startTimePresale;
+    startTimeSaleWithBonus = _startTimeSaleWithBonus;
+    startTimeSaleNoBonus = _startTimeSaleNoBonus;
+    endTime = _endTime;
+
+    require(_rate > 0);
+    rate = _rate;
+
+    require(_crowdSaleZxcCap > 0);
+    require(token.totalSupply() >= _crowdSaleZxcCap);
+
+    crowdSaleZxcCap = _crowdSaleZxcCap;
+    zxcSold = 0;
+
+    require(_minimumWeiDeposit > 0);
+    minimumWeiDeposit = _minimumWeiDeposit;
+  }
+
+  /**
+   * @dev Fallback function can be used to buy tokens.
+   */
+  function()
+    external
+    payable
+  {
+    buyTokens(msg.sender);
+  }
+
+  /**
+   * @dev Low level token purchase function.
+   * @param beneficiary Address which is buying tokens.
+   */
+  function buyTokens(address beneficiary)
+    public
+    payable
+  {
+    uint256 weiAmount = msg.value;
+
+    require(beneficiary != address(0));
+    require(weiAmount >= minimumWeiDeposit);
+
+    //NOTE: this reverts if NOT in any of the sale time periods!
+    uint256 tokens = getTokenAmount(weiAmount);
+
+    require(zxcSold.add(tokens) <= crowdSaleZxcCap);
+    zxcSold = zxcSold.add(tokens);
+
+    forwardFunds();
```

Comment:
Have checked the gas used by the `function(){}`? Because we have lots of logic here it could potentially cause problems. Ref: https://solidity.readthedocs.io/en/develop/contracts.html?highlight=fallback%20%202300#fallback-function

Reply:
Fallback function to purchase tokens used 93752 gas when testing with Ganache.